### PR TITLE
Remove a type restriction in Base.QuadGK.Segment (#19626)

### PR DESCRIPTION
**Pull request by ajkeller34**
See #19626

### CODE REVIEW

**Comment by tkelman** [comment]
is what you're trying to accomplish testable without needing the rest of Unitful.jl? there are some tests for quadgk in `test/math..jl`

**Comment by ajkeller34** [comment]
I have an idea for how to do this; will get back to you once implemented.

**Comment by ajkeller34** [comment]
I added some tests. If you're wondering why I used a module instead of putting my additions into the quadgk test set, it is because I was running into some namespace issue that caused the tests to fail. I put my new definitions and tests into a module like done [here](https://github.com/JuliaLang/julia/blob/3daa183c909211e132399bdfe1417b51437d5b3b/test/arrayops.jl#L1534) and the tests pass now.

File: test/math.jl [hunk 1]
```diff
@@ -760,6 +760,102 @@ end
...
+    function quadgk{T}(f, a::MockQuantity{T}, b::MockQuantity{T},
+        c::MockQuantity{T}...; abstol=zero(T), reltol=sqrt(eps(T)),
+        maxevals=10^7, order=7, norm=vecnorm)
+
+        _do_quadgk(f, [a, b, c...], order, T, abstol, reltol, maxevals, norm)
+    end
+
+    # Necessary with infinite or semi-infinite intervals since !(MockQuantity <: Real)
+    # and do_quadgk tests if eltype(s) <: Real.
+    function _do_quadgk{Tw,T<:Real}(f, s::Array{MockQuantity{T},1}, n, ::Type{Tw},
```
Comment (stevengj) [inline test/math.jl]: This seems overly complicated.   For testing this PR, it should be sufficient to have an `f(x)` where the *integrand* `f` is a `MockQuantity`, but `x` is still `Float64`.

File: test/math.jl [hunk 2]
```diff
@@ -760,6 +760,102 @@ end
...
+    using Base.Test
+    # Begin tests for 19626: Unitful compatibility.
+    # Define a mock physical quantity type
+    immutable MockQuantity{T} <: Number
+        val::T
+    end
+    Base.promote_rule{T,S}(::Type{MockQuantity{T}}, ::Type{MockQuantity{S}}) =
+        MockQuantity{promote_type(T,S)}
+    Base.convert{T}(::Type{MockQuantity{T}}, x::MockQuantity) = MockQuantity(T(x.val))
+
```
Comment (stevengj) [inline test/math.jl]: You shouldn't need to define promotion rules and conversion for a minimal working example.   Just define the minimum number of methods necessary for a test.

File: test/math.jl [hunk 3]
```diff
@@ -760,6 +760,102 @@ end
     @test quadgk(cos, 0,0.7,1, norm=abs)[1] ≈ sin(1)
 end
 
+module Test19626
+    using Base.Test
+    # Begin tests for 19626: Unitful compatibility.
+    # Define a mock physical quantity type
+    immutable MockQuantity{T} <: Number
```
Comment (stevengj) [inline test/math.jl]: It seems like it should work now even if it is not a subtype of `Number`?

File: test/math.jl (hunk 1 above)
Comment (ajkeller34) [inline test/math.jl]: Agreed, I became overexuberant thinking about how I would do this in Unitful.

File: test/math.jl (hunk 3 above)
Comment (ajkeller34) [inline test/math.jl]: If it is not a subtype of `Number` then I get a failure in `vecnorm`.

**Comment by ajkeller34** [comment]
I think I have the minimum number of definitions needed to get the test working now. Happy to implement further changes.

File: test/math.jl [hunk 4]
```diff
@@ -760,6 +760,35 @@ end
...
+    +(a::MockQuantity, b::MockQuantity) = MockQuantity(a.val+b.val)
+    -(a::MockQuantity, b::MockQuantity) = MockQuantity(a.val-b.val)
+    *(a::MockQuantity, b::Number) = MockQuantity(a.val*b)
+    abs(a::MockQuantity) = MockQuantity(abs(a.val))
+    isnan(a::MockQuantity) = isnan(a.val)
+    isinf(a::MockQuantity) = isinf(a.val)
+    isless(a::MockQuantity, b::MockQuantity) = isless(a.val, b.val)
+
+    # isless defn. necessary so that default abstol plays nicely with MockQuantity
+    isless(y::Number, x::MockQuantity) = y == 0 ? isless(MockQuantity(0), x) :
```
Comment (stevengj) [inline test/math.jl]: I would just pass `abstol = MockQuantity(0.0)`

File: test/math.jl [hunk 5]
```diff
@@ -760,6 +760,35 @@ end
     @test quadgk(cos, 0,0.7,1, norm=abs)[1] ≈ sin(1)
 end
 
+module Test19626
+    using Base.Test
+
+    # Define a mock physical quantity type
+    immutable MockQuantity{T} <: Number
+        val::T
```
Comment (stevengj) [inline test/math.jl]: Is the `<: Number` necessary?  I wouldn't bother parameterizing the type.  Just make `val::Float64`.

Comment (ajkeller34) [inline test/math.jl]: `<: Number` is necessary, otherwise I get an error from `vecnorm`.

**Comment by ajkeller34** [comment]
`<: Number` is necessary, otherwise I get an error from `vecnorm`. I have implemented the other changes.

**Review by stevengj — APPROVED** [review]

**Pull request merged**